### ISSUE TITLE
Support for sorting and filtering based upon relationship attributes

### ISSUE THREAD

**Issue by neomerx** [limoncello-php/json-api#33]
Original idea (in client application) https://github.com/neomerx/limoncello-flute/issues/17

First attempt was made to commit to base package https://github.com/neomerx/json-api/pull/161

Proper separation should be the following
- base package https://github.com/neomerx/json-api provides parsing capabilities for input filters though it does not convert them to any SQL queries as this part is very platform/framework specific. I do not expect any changes will be required in the base package while implementing this feature.
- actual transformation of query parameters to SQL via `Doctrine` is done in this package. A good starting point would be [applyFilters](https://github.com/limoncello-php/json-api/blob/master/src/Adapters/Repository.php#L303)

When the feature is implemented it most likely will be available in https://github.com/neomerx/limoncello-flute/ without any code changes.

**Plan**
- Develop tests for filtering results in belongsTo, hasMany and belongsToMany relationships. Inputs: filters, outputs: QueryBuilder (SQL) and (?) actual data from those queries. Examples are [here](https://github.com/limoncello-php/json-api/blob/master/tests/Adapters/RepositoryTest.php)
- Make those test work. I expect most if not all changes in `Repository.php` from reference above.

**Comment by neomerx** [comment]
A new branch `feature/issue33` was created for this feature

**Comment by neomerx** [comment]
@ngantingon I've added a test which will check that parsing of filters in relationships works. When code for the parser is added it will be able to reach `\Limoncello\JsonApi\Adapters\Repository::applyFilters` where actual transformation to SQL happens. Then tests for belongsTo, hasMany and belongsToMany filters could be added.

**Comment by neomerx** [comment]
@ngantingon I've added parsing (initial! but works for now) and now it's all about `\Limoncello\JsonApi\Adapters\Repository::applyFilters`

you can use `\Limoncello\Tests\JsonApi\Http\ControllerTest::testFilterBelongsToRelationship` to run the code and see in debugger how it works (not works atm :smile: ).

**Comment by neomerx** [comment]
@ngantingon I've added filtering for belongsTo and hasMany relationships. BelongsToMany will follow.

I'd appreciate your feedback and your style changes. Please post them to issue branch (not master).

Also I think sorting is not applicable as we do not return fields we would sort by.

**Comment by neomerx** [comment]
@ngantingon I've added filtering for belongsToMany as well. The result is fully covered with tests and ready to be merged. Please feel free to review and send your PR.

**Comment by ngantingon** [comment]
@neomerx yes. i did have bottleneck on sorting relationship attributes. the SortParameter was not designed for relationship attribute.

**Comment by ngantingon** [comment]
@neomerx added sorting on belongsTo relationship attributes and sent an PR for your review.

**Comment by neomerx** [comment]
@ngantingon PR has um-merged changes and fails in tests. Can you please run tests locally before posting? You can check it with `$ composer test`

Back to sorting. As I can see from the added test you are fetching comments with posts included and try to arrange those comments by `text` attribute of its post, right? Why do you include posts? Do you have any expectations about sorting order of those posts?

**Comment by ngantingon** [comment]
@neomerx having include post was to check if the sort direction follows as expected.
